File: Data Structures & Algorithms/max-area-of-island/submission-0.cs
public class Solution {
    public int MaxAreaOfIsland(int[][] grid) {
        var seen = new HashSet<(int, int)>();
        var maxArea = 0;

        var directions = new (int, int)[] {
            (0, 1),
            (1, 0),
            (0, -1),
            (-1, 0)
        };

        bool IsInvalidNode(int x, int y) {
            return x < 0 || x > grid.Length - 1 || y < 0 || y > grid[0].Length - 1;
        }

        int dfs(int x, int y) {
            if (IsInvalidNode(x, y) || grid[x][y] == 0 || seen.Contains((x,y))) return 0;

            var count = 1;
            seen.Add((x,y));

            foreach ((var dx, var dy) in directions) {
                count += dfs(x + dx, y + dy);
            }

            maxArea = Math.Max(maxArea, count);

            return count;
        }

        for (int i = 0; i < grid.Length; i++) {
            for (int j = 0; j < grid[0].Length; j++) {
                if (!seen.Contains((i,j))) {
                    dfs(i, j);
                }
            }
        }

        return maxArea;
    }
}

File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-0.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        seen = set()

        directions = [
            (1,1),
            (-1,-1),
            (1,-1),
            (-1,1),
            (0,-1),
            (0,1),
            (-1,0),
            (1,0),
        ]

        queue = deque()

        queue.append((0, 0, 1))
        shortest_path = math.inf

        while queue:
            r, c, count = queue.popleft()

            # base case
            if min(r,c) < 0 or max(r,c) >= len(grid) or grid[r][c] != 0:
                continue

            seen.add((r,c))

            if r == len(grid) - 1 and c == len(grid) - 1:
                shortest_path = min(shortest_path, count)

            for dr, dc in directions:
                if (r+dr, c+dc) not in seen:
                    queue.append((r+dr, c+dc, count + 1))

        if shortest_path != math.inf:
            return int(shortest_path)
        else:
            return -1
            